File: desktop/rask/charts.py
"""charts.py — Canvas-drawn charts (mirror of web/js/charts.js).

Renders on a tkinter.Canvas:
    ProgressRing.draw(canvas, cx, cy, size, progress, color, track, label, label_color)
    BarChart.draw(canvas, x, y, w, h, data, opts)
    DonutChart.draw(canvas, cx, cy, r, data, line_width)
    Heatmap.draw(canvas, x, y, w, h, year, data, cell_size)
"""
from __future__ import annotations
import colorsys
import datetime as _dt
import math
from typing import Dict, List, Optional, Tuple
from . import config


def hex_to_rgb(hex_str: str) -> Tuple[int, int, int]:
    h = (hex_str or config.GOLD).lstrip("#")
    if len(h) == 3:
        h = "".join(c + c for c in h)
    return int(h[0:2], 16), int(h[2:4], 16), int(h[4:6], 16)


def hex_to_rgba(hex_str: str, alpha: float = 1.0) -> str:
    r, g, b = hex_to_rgb(hex_str)
    # tkinter uses #RRGGBB; alpha is ignored by tkinter Canvas, but we expose it for API parity.
    return f"#{r:02x}{g:02x}{b:02x}"


# === ProgressRing (mirror of charts.js ProgressRing.draw) ===
def progress_ring(canvas, cx: float, cy: float, size: float, progress: float,
                  color: str = config.GOLD, track_color: str = config.SURFACE_HI,
                  label: str = "", label_color: str = config.TEXT) -> None:
    progress = max(0.0, min(1.0, progress))
    r = size / 2 - 8
    lw = 8
    # Track (full circle)
    canvas.create_oval(cx - r, cy - r, cx + r, cy + r,
                       outline=track_color, width=lw)
    # Progress arc (start at top, clockwise)
    if progress > 0:
        # Tkinter arc goes counterclockwise from start (degrees, east=0).
        # charts.js starts at -90° (top) and goes clockwise.
        # Tkinter arc: start=90 (top) and extent=-360*progress (clockwise = negative extent).
        extent = -360 * progress
        canvas.create_arc(cx - r, cy - r, cx + r, cy + r,
                          start=90, extent=extent, style="arc",
                          outline=color, width=lw)
    # Label
    if label:
        canvas.create_text(cx, cy, text=label, fill=label_color,
                           font=("Vazirmatn", max(10, int(size * 0.16)), "bold"))


# === BarChart (mirror of charts.js BarChart.draw) ===
def bar_chart(canvas, x: float, y: float, w: float, h: float,
              data: List[Dict[str, object]], opts: Optional[dict] = None) -> None:
    opts = opts or {}
    if not data:
        return
    values = [float(d.get("value", 0) or 0) for d in data]
    max_val = float(opts.get("maxValue") or max([1.0] + values))
    gap = 6
    bar_w = max(2.0, (w - gap * (len(data) + 1)) / len(data))
    base_y = y + h - 18
    # Baseline
    canvas.create_line(x, base_y, x + w, base_y, fill=config.SURFACE_HI, width=1)
    for i, d in enumerate(data):
        v = float(d.get("value", 0) or 0)
        bh = (v / max_val) * (h - 30) if max_val > 0 else 0
        bx = x + gap + i * (bar_w + gap)
        by = base_y - bh
        color = d.get("color") or config.GOLD
        # Rounded top (approximate with a rectangle — tkinter has no rounded rect natively)
        canvas.create_rectangle(bx, by, bx + bar_w, base_y, fill=color, outline="")
        # Label
        lbl = d.get("label")
        if lbl:
            canvas.create_text(bx + bar_w / 2, base_y + 6, text=str(lbl),
                               fill=config.TEXT_DIM, font=("Vazirmatn", 8))


# === DonutChart (mirror of charts.js DonutChart.draw) ===
def donut_chart(canvas, cx: float, cy: float, r: float,
                data: List[Dict[str, object]], line_width: int = 18) -> None:
    total = sum(float(d.get("value", 0) or 0) for d in data) or 1.0
    # Track
    canvas.create_oval(cx - r, cy - r, cx + r, cy + r,
                       outline=config.SURFACE_HI, width=line_width)
    angle = 90.0  # start at top
    for d in data:
        v = float(d.get("value", 0) or 0)
        if not v:
            continue
        seg = (v / total) * 360.0
        color = d.get("color") or config.GOLD
        # Clockwise = negative extent in tkinter
        canvas.create_arc(cx - r, cy - r, cx + r, cy + r,
                          start=angle, extent=-seg, style="arc",
                          outline=color, width=line_width)
        angle -= seg


# === Heatmap (mirror of charts.js Heatmap.draw) ===
def heatmap(canvas, x: float, y: float, w: float, h: float, year: int,
            data: Dict[str, int], cell_size: int = 12) -> None:
    gap = 3
    start = _dt.date(year, 1, 1)
    # Sunday of first week (mirror of charts.js: (getDay()+1)%7)
    py_weekday_to_js = lambda d: (d.weekday() + 1) % 7  # Mon=1, Sun=0, Sat=6
    offset = (py_weekday_to_js(start) + 1) % 7
    cursor = start - _dt.timedelta(days=offset)
    end = _dt.date(year, 12, 31)
    col = 0
    max_val = max([1] + list(data.values()))
    while cursor <= end:
        for row in range(7):
            if cursor.year == year:
                iso = cursor.isoformat()
                sec = data.get(iso, 0)
                tt = sec / max_val if max_val > 0 else 0
                color = _intensity_color(tt)
                cx0 = x + col * (cell_size + gap)
                cy0 = y + row * (cell_size + gap)
                canvas.create_rectangle(cx0, cy0, cx0 + cell_size, cy0 + cell_size,
                                        fill=color, outline="")
            cursor += _dt.timedelta(days=1)
        col += 1


def _intensity_color(t: float) -> str:
    steps = config.HEATMAP_STEPS
    idx = min(len(steps) - 1, int(t * len(steps)))
    r, g, b = steps[idx]
    return f"#{r:02x}{g:02x}{b:02x}"

File: web/js/charts.js
// charts.js — Custom canvas charts for Rask (no external chart library).
//   ProgressRing.draw(ctx, x, y, size, progress, color, trackColor, label)
//   BarChart.draw(ctx, x, y, w, h, data, opts)
//   DonutChart.draw(ctx, cx, cy, r, data, lineWidth)
//   Heatmap.draw(ctx, x, y, w, h, year, data, cellSize)
window.RaskCharts = (function () {
  function hexToRgba(hex, alpha) {
    hex = (hex || "#D4AF37").replace("#", "");
    if (hex.length === 3) hex = hex.split("").map((c) => c + c).join("");
    const r = parseInt(hex.slice(0, 2), 16);
    const g = parseInt(hex.slice(2, 4), 16);
    const b = parseInt(hex.slice(4, 6), 16);
    return `rgba(${r}, ${g}, ${b}, ${alpha == null ? 1 : alpha})`;
  }

  const ProgressRing = {
    draw(ctx, cx, cy, size, progress, color, trackColor, label, labelColor) {
      progress = Math.max(0, Math.min(1, progress));
      const r = size / 2 - 8;
      const lw = 8;
      ctx.clearRect(cx - size/2 - 2, cy - size/2 - 2, size + 4, size + 4);
      // Track
      ctx.beginPath();
      ctx.strokeStyle = trackColor || "#2C2C30";
      ctx.lineWidth = lw;
      ctx.arc(cx, cy, r, 0, Math.PI * 2);
      ctx.stroke();
      // Progress (start at top, go clockwise)
      if (progress > 0) {
        ctx.beginPath();
        ctx.strokeStyle = color || "#D4AF37";
        ctx.lineWidth = lw;
        ctx.lineCap = "round";
        ctx.arc(cx, cy, r, -Math.PI / 2, -Math.PI / 2 + Math.PI * 2 * progress);
        ctx.stroke();
      }
      // Label
      if (label) {
        ctx.fillStyle = labelColor || "#E8E8E8";
        ctx.font = `bold ${Math.floor(size * 0.16)}px Vazirmatn, sans-serif`;
        ctx.textAlign = "center";
        ctx.textBaseline = "middle";
        ctx.fillText(label, cx, cy);
      }
    }
  };

  const BarChart = {
    // data: [{label, value, color}]
    draw(ctx, x, y, w, h, data, opts) {
      opts = opts || {};
      ctx.clearRect(x, y, w, h);
      if (!data || !data.length) return;
      const maxVal = opts.maxValue || Math.max.apply(null, data.map((d) => d.value).concat([1]));
      const gap = 6;
      const barW = Math.max(2, (w - gap * (data.length + 1)) / data.length);
      // Baseline
      ctx.strokeStyle = "#2C2C30";
      ctx.lineWidth = 1;
      ctx.beginPath();
      ctx.moveTo(x, y + h - 18);
      ctx.lineTo(x + w, y + h - 18);
      ctx.stroke();
      // Bars
      data.forEach((d, i) => {
        const bh = maxVal > 0 ? (d.value / maxVal) * (h - 30) : 0;
        const bx = x + gap + i * (barW + gap);
        const by = y + h - 18 - bh;
        ctx.fillStyle = d.color || "#D4AF37";
        // Rounded top
        const radius = Math.min(3, barW / 2);
        ctx.beginPath();
        ctx.moveTo(bx, by + radius);
        ctx.arcTo(bx, by, bx + radius, by, radius);
        ctx.lineTo(bx + barW - radius, by);
        ctx.arcTo(bx + barW, by, bx + barW, by + radius, radius);
        ctx.lineTo(bx + barW, by + bh);
        ctx.lineTo(bx, by + bh);
        ctx.closePath();
        ctx.fill();
        // Label
        if (d.label) {
          ctx.fillStyle = "#9A9A9F";
          ctx.font = "10px Vazirmatn, sans-serif";
          ctx.textAlign = "center";
          ctx.textBaseline = "top";
          ctx.fillText(d.label, bx + barW / 2, y + h - 14);
        }
      });
    }
  };

  const DonutChart = {
    // data: [{label, value, color}]
    draw(ctx, cx, cy, r, data, lineWidth) {
      lineWidth = lineWidth || 18;
      ctx.save();
      // Clear area
      ctx.clearRect(cx - r - lineWidth, cy - r - lineWidth,
                    (r + lineWidth) * 2, (r + lineWidth) * 2);
      const total = data.reduce((s, d) => s + d.value, 0) || 1;
      // Track
      ctx.beginPath();
      ctx.strokeStyle = "#2C2C30";
      ctx.lineWidth = lineWidth;
      ctx.arc(cx, cy, r, 0, Math.PI * 2);
      ctx.stroke();
      // Segments
      let angle = -Math.PI / 2;
      data.forEach((d) => {
        if (!d.value) return;
        const seg = (d.value / total) * Math.PI * 2;
        ctx.beginPath();
        ctx.strokeStyle = d.color || "#D4AF37";
        ctx.lineWidth = lineWidth;
        ctx.lineCap = "butt";
        ctx.arc(cx, cy, r, angle, angle + seg);
        ctx.stroke();
        angle += seg;
      });
      ctx.restore();
    }
  };

  const Heatmap = {
    // data: { "YYYY-MM-DD": seconds }
    draw(ctx, x, y, w, h, year, data, cellSize) {
      cellSize = cellSize || 12;
      const gap = 3;
      ctx.clearRect(x, y, w, h);
      const max = Math.max.apply(null, Object.values(data).concat([1]));
      const start = new Date(year, 0, 1);
      // Find Sunday of first week
      const offset = (start.getDay() + 1) % 7;
      const cursor = new Date(start);
      cursor.setDate(cursor.getDate() - offset);
      const end = new Date(year, 11, 31);
      let col = 0;
      while (cursor <= end) {
        for (let row = 0; row < 7; row++) {
          if (cursor.getFullYear() === year) {
            const iso = cursor.toISOString().slice(0, 10);
            const sec = data[iso] || 0;
            const t = max > 0 ? sec / max : 0;
            const color = Heatmap.intensityColor(t);
            ctx.fillStyle = color;
            ctx.fillRect(
              x + col * (cellSize + gap),
              y + row * (cellSize + gap),
              cellSize, cellSize
            );
          }
          cursor.setDate(cursor.getDate() + 1);
        }
        col++;
      }
    },
    intensityColor(t) {
      const steps = [
        "rgba(14,14,16,1)",
        "rgba(48,40,16,1)",
        "rgba(80,64,20,1)",
        "rgba(120,96,28,1)",
        "rgba(212,175,55,1)",
      ];
      const idx = Math.min(steps.length - 1, Math.floor(t * steps.length));
      return steps[idx];
    }
  };

  return { hexToRgba, ProgressRing, BarChart, DonutChart, Heatmap };
})();
